File: scratch_db/extract_snappy.py
import os
import snappy
import sys

appdata = os.environ.get('APPDATA')
ls_path = os.path.join(appdata, 'react-example', 'Local Storage', 'leveldb')

all_data = b""
for f in os.listdir(ls_path):
    if f.endswith('.ldb'): # only check ldb files, log files might not have block compression
        with open(os.path.join(ls_path, f), 'rb') as file:
            data = file.read()
            if b"3derp_printJobs" in data:
                print(f"Found in {f}")
                pos = data.find(b"3derp_printJobs")
                # Try all block starts before this position
                for start in range(max(0, pos - 4096), pos + 1):
                    for length in range(100, min(8192, len(data) - start)):
                        try:
                            chunk = data[start:start+length]
                            uncompressed = snappy.uncompress(chunk)
                            if b"3derp_printJobs" in uncompressed and b"[{" in uncompressed:
                                print(f"Success at start {start}, length {length}")
                                with open('extracted_prints_raw.bin', 'wb') as out:
                                    out.write(uncompressed)
                                
                                # strip non-printable
                                clean_str = uncompressed.decode('utf-16le', errors='ignore')
                                import re
                                clean_str = re.sub(r'[\x00-\x1F\x7F]', '', clean_str)
                                with open('extracted_prints_clean.txt', 'w', encoding='utf-8') as out_txt:
                                    out_txt.write(clean_str)
                                
                                print("Wrote output files.")
                                sys.exit(0)
                        except Exception as e:
                            pass
                print("Failed to decompress block in this file")

File: scratch_db/extract_snappy.js
const fs = require('fs');
const path = require('path');
const snappy = require('snappy');

const lsPath = path.join(process.env.APPDATA, 'react-example', 'Local Storage', 'leveldb');
const files = fs.readdirSync(lsPath).filter(f => f.endsWith('.ldb') || f.endsWith('.log'));

let allData = [];
files.forEach(f => {
  try {
    allData.push(fs.readFileSync(path.join(lsPath, f)));
  } catch(e){}
});
const data = Buffer.concat(allData);

const target = Buffer.from('3derp_printJobs');

let pos = 0;
while (true) {
  pos = data.indexOf(target, pos);
  if (pos === -1) break;
  
  console.log('Found 3derp_printJobs at pos', pos);
  
  // Try decompressing from here or slightly after
  for (let offset = 15; offset < 40; offset++) {
    for (let len = 10; len < 5000; len++) {
      try {
        const slice = data.slice(pos + offset, pos + offset + len);
        const uncompressed = snappy.uncompressSync(slice);
        const str = uncompressed.toString('utf8');
        if (str.includes('[{"')) {
          console.log(`Success at offset ${offset}, len ${len}!`);
          fs.writeFileSync('extracted_prints_raw.bin', uncompressed);
          const cleanStr = uncompressed.toString('utf16le').replace(/[\x00-\x1F\x7F]/g, '');
          fs.writeFileSync('extracted_prints_clean.txt', cleanStr);
          console.log('Wrote to extracted_prints_clean.txt');
          process.exit(0);
        }
      } catch (e) {
        // ignore
      }
    }
  }
  
  pos += target.length;
}
console.log('Done scanning.');
